File: advall-haskell/d21.hs
import Data.List

parseInput :: String -> [(String, String, String)]
parseInput s = map parseInput' (map words (lines s))

parseInput' :: [String] -> (String, String, String)
parseInput' ("swap":"position":xs) = ("swap", head xs, last xs)
parseInput' ("swap":"letter":xs) = ("swapLet", head xs, last xs)
parseInput' ("rotate":"based":xs) = ("rotatePos", last xs, "")
parseInput' ("rotate":dir:xs) = ("rotate", dir, head xs)
parseInput' ("reverse":_:xs) = ("reverse", head xs, last xs)
parseInput' ("move":_:xs) = ("move", head xs, last xs)

scramble :: String -> [(String, String, String)] -> String
scramble s [] = s
scramble s ((i,p1,p2):is)
    | i == "swap" = scramble (swap s pi1 pi2) is
    | i == "swapLet" = scramble (swap s (head (findIndices (==(head p1)) s)) (head (findIndices (==(head p2)) s))) is
    | i == "rotatePos" = scramble (rotateB s (head p1)) is
    | i == "rotate" = scramble (rotate s p1 pi2) is
    | i == "reverse" = scramble (revT s pi1 pi2) is
    | i == "move" = scramble (mv s pi1 pi2) is
{-    | i == "swap" = scramble (swap s pi1 pi2) is
    | i == "swapLet" = scramble s is
    | i == "rotatePos" = scramble (rotateB s (head p1)) is
    | i == "rotate" = scramble (rotate s p1 pi2) is
    | i == "reverse" = scramble (revT s pi1 pi2) is
    | i == "move" = scramble (mv s pi1 pi2) is-}
    where
        pi1 = read p1 :: Int
        pi2 = read p2 :: Int

swap :: String -> Int -> Int -> String
swap s x y
    | x == y = s
    | x > y = swap s y x
    | otherwise = (take x s) ++ [s!!y] ++ (drop (x+1) (take y s)) ++ [s!!x] ++ (drop (y+1) s)

rotate :: String -> String -> Int -> String
rotate s dir x
    | x == 0 = s
    | dir == "right" = rotate ([last s] ++ (init s)) dir (x-1)
    | dir == "left" = rotate ((tail s) ++ [head s]) dir (x-1)

rotateB :: String -> Char -> String
rotateB s x 
    | (head (findIndices (==x) s)) > 3 = rotate s "right" (2 + (head (findIndices (==x) s)))
    | otherwise = rotate s "right" (1 + (head (findIndices (==x) s)))

revT :: String -> Int -> Int -> String
revT s x y = (take x s) ++ (reverse (drop x (take (y+1) s))) ++ (drop (y+1) s)

mv :: String -> Int -> Int -> String
mv s x y
    | x <= y = (take x s) ++ (drop (x+1) (take (y+1) s)) ++ [s!!x] ++ (drop (y+1) s)
    | x > y = (take y s) ++ [s!!x] ++ (drop (y) (take x s)) ++ (drop (x+1) s)

unscramble :: String -> [(String, String, String)] -> [String] -> String
unscramble scrambled is (c:cs)
    | (scramble c is) == scrambled = c
    | otherwise = unscramble scrambled is cs

main = do
    input <- readFile "d21_input.txt"
    let answer1 = scramble "abcdefgh" (parseInput input)
    putStrLn $ "Part one: " ++ answer1
    let answer2 = unscramble "fbgdceah" (parseInput input) (permutations "fbgdceah")
    putStrLn $ "Part two: " ++ answer2

File: advall-python/d21.py
from itertools import permutations

inputPath = "d21_input.txt"

def swap(s, x, y):
    xVal = s[x]
    yVal = s[y]
    s = s.replace(xVal, "x")
    s = s.replace(yVal, "y")
    s = s.replace("x", yVal)
    s = s.replace("y", xVal)
    return s

def rotate(s, x):
    steps = abs(x) % len(s)
    if x >= 0:
        steps = steps * (-1)
    return (s[(steps):]) + (s[0:steps])

def revT(s, x, y):
    reversedPart = s[x:(y+1)]
    reversedPart = reversedPart[::-1]
    return s[0:x] + reversedPart + s[(y+1):]

def mv(s, x, y):
    if x == y:
        return s
    xVal = s[x]
    yVal = s[y]
    s = s.replace(xVal, "")
    if y < x:
        s = s.replace(yVal, (xVal + yVal))
    else: 
        s = s.replace(yVal, (yVal + xVal))
    return s

def scramble(s, inss):
    pwd = s
    for ins in inss:
        if (ins[0] == "swap") and (ins[1] == "position"):
            pwd = swap(pwd, int(ins[2]), int(ins[5]))
        elif (ins[0] == "swap") and (ins[1] == "letter"):
            pwd = swap(pwd, pwd.index(ins[2]), pwd.index(ins[5]))
        elif (ins[0] == "rotate") and (ins[1] == "right"):
            pwd = rotate(pwd, int(ins[2]))
        elif (ins[0] == "rotate") and (ins[1] == "left"):
            pwd = rotate(pwd, (int(ins[2]) * (-1)))
        elif (ins[0] == "rotate") and (ins[1] == "based"):
            steps = pwd.index(ins[6]) + 1
            if steps > 4:
                steps += 1
            pwd = rotate(pwd, steps)
        elif ins[0] == "reverse":
            pwd = revT(pwd, int(ins[2]), int(ins[4]))
        elif ins[0] == "move":
            pwd = mv(pwd, int(ins[2]), int(ins[5]))
    return pwd

def parseInput(path):
    with open (path, "r") as inputfile:
        input = inputfile.readlines()
    lines = []
    for line in input:
        lines.append(line.split())
    return lines

def unscramble(s, inss):
    candidates = ["".join(p) for p in permutations(s)]
    for cand in candidates:
        if s == scramble(cand, inss):
            return cand
    print "No solution to part 2 was found!"

print "Part one: " + scramble("abcdefgh", parseInput(inputPath))
print "Part two: " + unscramble("fbgdceah", parseInput(inputPath))
